cpp/tests/groupby/group_product_test.cpp

```diff
@@ -0,0 +1,149 @@
+}
+
+TYPED_TEST(groupby_product_test, empty_cols)
+{
+    using K = int32_t;
+    using V = TypeParam;
+    using R = cudf::detail::target_type_t<V, aggregation::PRODUCT>;
+
+    fixed_width_column_wrapper<K> keys        { };
+    fixed_width_column_wrapper<V> vals        { };
+
+    fixed_width_column_wrapper<K> expect_keys { };
+    fixed_width_column_wrapper<R> expect_vals { };
+
+    auto agg = cudf::make_product_aggregation();
+    test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg));
+}
+
+TYPED_TEST(groupby_product_test, zero_valid_keys)
+{
+    using K = int32_t;
+    using V = TypeParam;
+    using R = cudf::detail::target_type_t<V, aggregation::PRODUCT>;
+
+    fixed_width_column_wrapper<K> keys      ( { 1, 2, 3}, all_null() );
+    fixed_width_column_wrapper<V> vals        { 3, 4, 5};
+
+    fixed_width_column_wrapper<K> expect_keys { };
+    fixed_width_column_wrapper<R> expect_vals { };
+
+    auto agg = cudf::make_product_aggregation();
+    test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg));
+}
+
+TYPED_TEST(groupby_product_test, zero_valid_values)
+{
+    using K = int32_t;
+    using V = TypeParam;
+    using R = cudf::detail::target_type_t<V, aggregation::PRODUCT>;
+
+    fixed_width_column_wrapper<K> keys        { 1, 1, 1};
+    fixed_width_column_wrapper<V> vals      ( { 3, 4, 5}, all_null() );
+
+    fixed_width_column_wrapper<K> expect_keys { 1 };
+    fixed_width_column_wrapper<R> expect_vals({ 0 }, all_null());
+
+    auto agg = cudf::make_product_aggregation();
+    test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg));
+}
+
+TYPED_TEST(groupby_product_test, null_keys_and_values)
+{
+    using K = int32_t;
+    using V = TypeParam;
+    using R = cudf::detail::target_type_t<V, aggregation::PRODUCT>;
+
+    fixed_width_column_wrapper<K> keys(       { 1, 2, 3, 1, 2, 2, 1, 3, 3, 2, 4},
+                                              { 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1});
+    fixed_width_column_wrapper<V> vals(       { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 3},
+                                              { 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0});
+
+                                          //  { 1, 1,     2, 2, 2,   3, 3,    4}
+    fixed_width_column_wrapper<K> expect_keys({ 1,        2,         3,       4}, all_valid());
+                                          //  { _, 3, 6,  1, 4, 9,   2, 8,    _}
+    fixed_width_column_wrapper<R> expect_vals({ 18.,      36.,       16.,     3.},
+                                              { 1,        1,         1,       0});
+
+    auto agg = cudf::make_product_aggregation();
+    test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg));
+}
+// clang-format on
+
+// This test will not work until the following ptxas bug is fixed in 10.2
+// https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317&cp=
+TYPED_TEST(groupby_product_test, DISABLED_dictionary)
```

Comment:
```suggestion
TYPED_TEST(groupby_product_test, dictionary)
```

Reply:
This may need another cleanup. At current state, code needs specialization for every keys type. 
Also another PR #7949 is also adding few operations.

Reply:
OK, if you think it's best to wait to enable this test feel free to resolve.